src/Compilers/CSharp/Portable/Binder/Binder_Statements.cs

```diff
@@ -1724,104 +1724,300 @@ private BoundExpression BindAssignment(AssignmentExpressionSyntax node, Diagnost
+                {
+                    var nestedArguments = ((TupleExpressionSyntax)argument.Expression).Arguments;
+                    checkedVariablesBuilder.Add(new BoundDeconstructionVariables(argument, BindDeconstructionVariables(nestedArguments, diagnostics)));
+                }
+                else
+                {
+                    var boundVariable = BindExpression(argument.Expression, diagnostics, invoked: false, indexed: false);
+                    var checkedVariable = CheckValue(boundVariable, BindValueKind.Assignment, diagnostics);
+
+                    checkedVariablesBuilder.Add(checkedVariable);
+                }
+            }
+
+            var checkedVariables = checkedVariablesBuilder.ToImmutableAndFree();
+            return checkedVariables;
         }
 
         /// <summary>
         /// Figures out how to assign from sourceType into receivingVariable and bundles the information (leaving holes for the actual source and receiver) into an AssignmentInfo.
         /// </summary>
-        private BoundDeconstructionAssignmentOperator.AssignmentInfo MakeAssignmentInfo(BoundExpression receivingVariable, TypeSymbol sourceType, AssignmentExpressionSyntax node, DiagnosticBag diagnostics)
+        private BoundDeconstructionAssignmentOperator.AssignmentStep MakeAssignmentInfo(BoundExpression receivingVariable, TypeSymbol sourceType, BoundDeconstructValuePlaceholder rightPlaceholder, AssignmentExpressionSyntax node, DiagnosticBag diagnostics)
         {
-            var leftPlaceholder = new BoundLValuePlaceholder(receivingVariable.Syntax, receivingVariable.Type) { WasCompilerGenerated = true };
-            var rightPlaceholder = new BoundRValuePlaceholder(node.Right, sourceType) { WasCompilerGenerated = true };
+            var leftPlaceholder = new BoundDeconstructValuePlaceholder(receivingVariable.Syntax, receivingVariable.Type) { WasCompilerGenerated = true };
 
             // each assignment has a placeholder for a receiver and another for the source
             BoundAssignmentOperator op = BindAssignment(node, leftPlaceholder, rightPlaceholder, diagnostics);
 
-            return new BoundDeconstructionAssignmentOperator.AssignmentInfo() { Assignment = op, LValuePlaceholder = leftPlaceholder, RValuePlaceholder = rightPlaceholder };
+            return new BoundDeconstructionAssignmentOperator.AssignmentStep() { Assignment = op, OutputPlaceholder = leftPlaceholder, TargetPlaceholder = rightPlaceholder };
+        }
+
+        private ImmutableArray<BoundExpression> FlattenDeconstructVariables(ImmutableArray<BoundExpression> variables)
+        {
+            bool isFlat = true;
+            foreach (var variable in variables)
+            {
+                if (variable.Kind == BoundKind.DeconstructionVariables) { isFlat = false; break; }
+            }
+
+            if (isFlat)
+            {
+                return variables;
+            }
```

Comment:
Done. I was nervous about using LINQ, but now I understand the relevant factor (does it capture local variables?).